### PR TITLE
Determinisitically get kubectl proxy port

### LINKED ISSUE #19252: integration flake: Timed out waiting for kubectl proxy
It seems like an issue similar to the problem with API server. I'll extend the timeout in the similar fashion.

cc @thockin @ixdy

### PR DESCRIPTION
**Pull request by lavalamp**
Fixes #19252

### CODE REVIEW

File: hack/test-cmd.sh [hunk 1]
```diff
@@ -38,22 +41,23 @@ function start-proxy()
...
-    if kube::util::test_host_port_free "127.0.0.1" "${PROXY_PORT}"; then
-      if [ $# -eq 0 ]; then
-        kubectl proxy -p ${PROXY_PORT} --www=. 1>&2 & break
-      else
-        kubectl proxy -p ${PROXY_PORT} --www=. --api-prefix="$1" 1>&2 & break
-      fi
-    fi
-    sleep 1;
+  if [ $# -eq 0 ]; then
+    kubectl proxy --port=0 --www=. 1>proxy-port.out &
```
Comment (ixdy) [inline hack/test-cmd.sh]: wonder if we should use `mktemp` so there's no chance of colliding with another parallel run?

File: hack/test-cmd.sh [hunk 2]
```diff
@@ -31,6 +31,9 @@ function stop-proxy()
   [[ -n "${PROXY_PID-}" ]] && kill "${PROXY_PID}" 1>&2 2>/dev/null
   PROXY_PID=
   PROXY_PORT=
+
+  # Delete file we used to store the proxy's output (to get the port).
+  rm proxy-port.out
```
Comment (ixdy) [inline hack/test-cmd.sh]: maybe use `rm -f` so this won't errexit out if this file doesn't exist

File: hack/test-cmd.sh [hunk 3]
```diff
@@ -38,22 +41,23 @@ function start-proxy()
...
-    sleep 1;
+  if [ $# -eq 0 ]; then
+    kubectl proxy --port=0 --www=. 1>proxy-port.out &
+  else
+    kubectl proxy --port=0 --www=. --api-prefix="$1" 1>proxy-port.out &
+  fi
+  PROXY_PID=$!
+  PROXY_PORT=
+
+  while [[ -z ${PROXY_PORT} ]]; do
```
Comment (ixdy) [inline hack/test-cmd.sh]: maybe we want a timeout in case the proxy never starts?

**Comment by lavalamp** [comment]
@ixdy Thanks, I think I addressed your comments. PTAL

**Comment by lavalamp** [comment]
Rebased and pushed, I guess docs were wrong at head briefly?

File: hack/test-cmd.sh [hunk 4]
```diff
@@ -38,22 +41,32 @@ function start-proxy()
...
+  else
+    kubectl proxy --port=0 --www=. --api-prefix="$1" 1>${PROXY_PORT_FILE} 2>&1 &
+  fi
+  PROXY_PID=$!
+  PROXY_PORT=
+
+  local attempts=0
+  while [[ -z ${PROXY_PORT} ]]; do
+    if (( ${attempts} > 9 )); then
+      kube::log::status "Couldn't start proxy. Failed to read port after ${attempts} tries. Got: $(cat ${PROXY_PORT_FILE})"
```
Comment (ixdy) [inline hack/test-cmd.sh]: maybe use `kube::log::error_exit`?
also maybe we want to `kill "${PROXY_PID}"`?

**Comment by lavalamp** [comment]
Fixed, PTAL

**Comment by ixdy** [comment]
LGTM

**Comment by lavalamp** [comment]
Thanks!

@k8s-oncall this is a flake fix, please merge manually

**Comment by fabioy** [comment]
@lavalamp SG. I'm waiting for a e2e green and I'll manually merge this one. Thanks.

**Pull request merged**